### ISSUE TITLE
Question: no valid interleaving, but why?

### ISSUE THREAD

**Issue by dnadales** [hedgehogqa/haskell-hedgehog#221]
I've made a [couple of implementations](https://github.com/dnadales/sandbox/tree/d9ebe1cb6f0f0cea72eb4ec8448ebae3255313b7/hedgehog-sm-echo) of a system with the following API:

```haskell
class Echo e where
    -- | Input a string. Returns 'True' iff the buffer was empty and the given
    -- string was added to it.
    input :: e -> String -> IO Bool

    -- | Output the buffer contents (if any).
    output :: e -> IO (Maybe String)

    -- | Reset the echo state.
    reset :: e -> IO ()
```

[One of the implementations](https://github.com/dnadales/sandbox/blob/d9ebe1cb6f0f0cea72eb4ec8448ebae3255313b7/hedgehog-sm-echo/src/EchoTVarWrong.hs) is wrongly implemented, and the `hedgehog` [specification](https://github.com/dnadales/sandbox/blob/d9ebe1cb6f0f0cea72eb4ec8448ebae3255313b7/hedgehog-sm-echo/test/EchoSpec.hs) detects this:

```text
━━━ EchoSpec ━━━
  ✓ prop_echoTVar passed 100 tests.
  ✓ prop_echoTVarPar passed 100 tests.
  ✓ prop_echoMVar passed 100 tests.
  ✓ prop_echoMVarPar passed 100 tests.
  ✓ prop_echoTVarWrong passed 100 tests.
  ✗ prop_echoTVarParWrong failed after 5 tests.

        ┏━━ test/EchoSpec.hs ━━━
    105 ┃ prop_echo_par :: Echo e => IO e -> Property
    106 ┃ prop_echo_par mkEcho = property $ do
    107 ┃     env <- liftIO mkEcho
    108 ┃     actions <- forAll $
    109 ┃         Gen.parallel (Range.linear 1 100)
    110 ┃                      (Range.linear 1 12)
    111 ┃                      initialState
    112 ┃                      [ cInput env
    113 ┃                      , cOutput env
    114 ┃                      ]
        ┃                      │ ━━━ Prefix ━━━
        ┃                      │ Var 0 = In "buzz"
        ┃                      │ Var 1 = Out
        ┃                      │
        ┃                      │ ━━━ Branch 1 ━━━
        ┃                      │ Var 2 = In "baz"
        ┃                      │
        ┃                      │ ━━━ Branch 2 ━━━
        ┃                      │ Var 3 = In "bar"
    115 ┃     liftIO $ reset env
    116 ┃     test $ executeParallel initialState actions

    no valid interleaving
```
What I understand is that no valid interleaving means that there's no sequential version of the counterexample above that explains the results that were observed. However I'd like to know which where these results. Can `hedgehog` offer some additional information?

What I assume is that both branches got the result "True" from the system-under-test (meaning that the string was successfully added to the buffer of size one). But if so, does that mean that after the prefix both branches have a copy of the model?

**Comment by HuwCampbell** [comment]
G'day, sorry for the issue lag.

Yep, your assumptions are correct. No valid interleaving means that there's no interleaved ordering which the actions in branches 1 and 2 can be taken which gives a result consistent with the output received when running them in parallel.

You're assumption that both return True is almost certainly right as well. With regards to the last statement, regarding both branches having a copy of the model, that's not so clear. Both branches use their symbolic models (starting at the prefix) to generate the action sequences, just the actions are then run in parallel; then finally, all interleavings are tested with a sequential model.

> Can hedgehog offer some additional information?

I believe that's it's plausible we could add the values received for each `Var` as an annotation to `executeParallel`. I have been prototyping an implementation, but it requires an extra constraint or two, so the maintainers would need to weigh up the pros and cons.

**Comment by HuwCampbell** [comment]
My work on this was merged in #235. Should this issue be closed?

**Comment by dnadales** [comment]
Yes! Thank you so much for this @HuwCampbell !
